tools/state-viewer/src/contract_accounts.rs

```diff
@@ -0,0 +1,164 @@
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ContractAccountError {
+    #[error("could not parse key {1:?}")]
+    InvalidKey(#[source] std::io::Error, Vec<u8>),
+    #[error("failed loading contract code for account {1}")]
+    NoCode(#[source] StorageError, AccountId),
+}
+
+impl std::fmt::Display for ContractAccount {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:<64} {:>9}", self.account_id, self.source_wasm.len())
+    }
+}
+
+impl ContractAccount {
+    /// Iterate over all contracts stored in the given trie, in lexicographic
+    /// order of the account IDs.
+    pub(crate) fn in_trie(trie: &Trie) -> anyhow::Result<ContractAccountIterator> {
+        ContractAccountIterator::new(trie)
+    }
+
+    fn from_contract_trie_node(
+        trie_key: &[u8],
+        value_hash: CryptoHash,
+        trie: &Trie,
+    ) -> Result<Self, ContractAccountError> {
+        let account_id = parse_account_id_from_contract_code_key(trie_key)
+            .map_err(|err| ContractAccountError::InvalidKey(err, trie_key.to_vec()))?;
+        let source_wasm = trie
+            .storage
+            .retrieve_raw_bytes(&value_hash)
+            .map_err(|err| ContractAccountError::NoCode(err, account_id.clone()))?;
+        Ok(Self { account_id, source_wasm })
+    }
+}
+
+pub(crate) struct ContractAccountIterator<'a> {
+    /// Trie nodes that point to the contracts.
+    contract_nodes: VecDeque<TrieTraversalItem>,
+    trie: &'a Trie,
+}
+
+impl<'a> ContractAccountIterator<'a> {
+    pub(crate) fn new(trie: &'a Trie) -> anyhow::Result<Self> {
+        let mut trie_iter = trie.iter()?;
+        // `TrieKey::ContractCode` requires a valid `AccountId`, we use "xx"
+        let key = TrieKey::ContractCode { account_id: "xx".parse()? }.to_vec();
+        let (prefix, suffix) = key.split_at(key.len() - 2);
+        assert_eq!(suffix, "xx".as_bytes());
+
+        // `visit_nodes_interval` wants nibbles stored in `Vec<u8>` as input
+        let nibbles_before: Vec<u8> = NibbleSlice::new(prefix).iter().collect();
+        let nibbles_after = {
+            let mut tmp = nibbles_before.clone();
+            *tmp.last_mut().unwrap() += 1;
+            tmp
+        };
+
+        // finally, use trie iterator to find all contract nodes
+        let vec_of_nodes = trie_iter.visit_nodes_interval(&nibbles_before, &nibbles_after)?;
+        let contract_nodes = VecDeque::from(vec_of_nodes);
+        Ok(Self { contract_nodes, trie })
+    }
+}
+
+impl Iterator for ContractAccountIterator<'_> {
+    type Item = Result<ContractAccount, ContractAccountError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while !self.contract_nodes.is_empty() {
+            // only look at nodes with a value, ignoring intermediate nodes
+            // without values
+            if let Some(TrieTraversalItem { hash, key: Some(trie_key) }) =
+                self.contract_nodes.pop_front()
```

Comment:
optional suggestion/nit:
```
while let Some(item) = self.contract_nodes.pop_front() {
    if let Some(TrieTraversalItem { hash, key: Some(trie_key) }) = item {
       ...
    }
}
```
To me this seems to be a bit more easier to read, no strong opinion though, up to you :)

Reply:
oh yeah, I like it, thanks!